### PR TITLE
Read Identifiers for Search Api from Dynamics

### PR DESCRIPTION
**Pull request by peggy-gov-bc-ca**
# Description
Read Identifiers for Search Api from Dynamics

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
tested on my local pc

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

## For bcgov contributors:

this PR fixes jira ticket: FAMS-737

### CODE REVIEW

File: app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs [hunk 1]
```diff
@@ -38,13 +38,39 @@ public SearchApiRequestService(IODataClient oDataClient)
...
+
+                IEnumerable<SSG_SearchApiRequest> searchApiRequests = await _oDataClient.For<SSG_SearchApiRequest>()
+                    .Select(x => x.SearchApiRequestId)
+                    .Filter(x => x.StatusCode == readyForSearchCode)
+                    .FindEntriesAsync(cancellationToken);
+
+                List<SSG_SearchApiRequest> results = new List<SSG_SearchApiRequest>();
+                foreach (SSG_SearchApiRequest request in searchApiRequests)
+                {
+                    SSG_SearchApiRequest r = await _oDataClient.For<SSG_SearchApiRequest>()
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs]: ok. Actually, when the variable life scope is very short, I like to use very short name to indicate it is just a temp thing. 
I will change it to a longer name.

File: app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs [hunk 2]
```diff
@@ -38,13 +38,39 @@ public SearchApiRequestService(IODataClient oDataClient)
...
-            int statusCode = SearchApiRequestStatusReason.ReadyForSearch.Value;
-            return await _oDataClient.For<SSG_SearchApiRequest>()
-                .Filter(x => x.StatusCode == statusCode)
-                .FindEntriesAsync(cancellationToken);
+            int readyForSearchCode = SearchApiRequestStatusReason.ReadyForSearch.Value;
+
+            try
+            {
+
+                //IEnumerable<SSG_SearchApiRequest> results = await _oDataClient.For<SSG_SearchApiRequest>().Filter(x => x.StatusCode == readyForSeachCode).Expand(x => x.Identifiers)
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs]: ok

File: app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs [hunk 3]
```diff
@@ -38,13 +38,39 @@ public SearchApiRequestService(IODataClient oDataClient)
...
+                    SSG_SearchApiRequest r = await _oDataClient.For<SSG_SearchApiRequest>()
+                        .Key(request.SearchApiRequestId)
+                        .Expand(x => x.Identifiers).FindEntryAsync(cancellationToken);
+                    results.Add(r);
+                }
+                return results;
+            }
+            catch (Exception ex)
+            {
+                throw;
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs]: ok. will remove

File: app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs [hunk 4]
```diff
@@ -38,13 +38,39 @@ public SearchApiRequestService(IODataClient oDataClient)
...
+
+                //todo: we need to change to use above code, but ODataClient 4 has problems with expand, following code is a workaround
+                //ref: https://powerusers.microsoft.com/t5/Power-Apps-Ideas/Web-API-Implement-expand-on-collections/idi-p/221291
+
+                IEnumerable<SSG_SearchApiRequest> searchApiRequests = await _oDataClient.For<SSG_SearchApiRequest>()
+                    .Select(x => x.SearchApiRequestId)
+                    .Filter(x => x.StatusCode == readyForSearchCode)
+                    .FindEntriesAsync(cancellationToken);
+
+                List<SSG_SearchApiRequest> results = new List<SSG_SearchApiRequest>();
```
[the remark this answers is not in the document]
Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/Fams3Adapter.Dynamics/SearchApiRequest/SearchApiRequestService.cs]: yep

File: app/DynamicsAdapter/Fams3Adapter.Dynamics.Test/SearchApiRequest/SearchApiRequestServiceTest.cs [hunk 5]
```diff
@@ -29,16 +29,50 @@ public void SetUp()
             int readyForSearchVAlue = SearchApiRequestStatusReason.ReadyForSearch.Value;
             int inProgressValue = SearchApiRequestStatusReason.InProgress.Value;
 
+            //odataClientMock.Setup(x => x.For<SSG_SearchApiRequest>(null)
```
Comment (alexjoybc) [inline app/DynamicsAdapter/Fams3Adapter.Dynamics.Test/SearchApiRequest/SearchApiRequestServiceTest.cs]: Remove commented code ;-)

File: app/DynamicsAdapter/Fams3Adapter.Dynamics/Identifier/InformationSourceType.cs [hunk 6]
```diff
@@ -0,0 +1,17 @@
+﻿using Fams3Adapter.Dynamics.OptionSets.Models;
+
+namespace Fams3Adapter.Dynamics.Identifier
+{
+    public class InformationSourceType : Enumeration
+    {
+
+        public static InformationSourceType DriverLicense = new InformationSourceType(867670000, "Request");
```
Comment (alexjoybc) [inline app/DynamicsAdapter/Fams3Adapter.Dynamics/Identifier/InformationSourceType.cs]: why DriverLicence = Request ?

File: app/DynamicsAdapter/Fams3Adapter.Dynamics.Test/SearchApiRequest/SearchApiRequestServiceTest.cs (hunk 5 above)
Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/Fams3Adapter.Dynamics.Test/SearchApiRequest/SearchApiRequestServiceTest.cs]: good catch. sorry to forget to remove it.

File: app/DynamicsAdapter/Fams3Adapter.Dynamics/Identifier/InformationSourceType.cs (hunk 6 above)
Comment (peggy-gov-bc-ca) [reply app/DynamicsAdapter/Fams3Adapter.Dynamics/Identifier/InformationSourceType.cs]: good catch. copy paste, then forget to change it. :(

**Pull request merged**